### PR TITLE
[FLINK-19924][config] Disallow UC for iterative jobs

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

From 1.12, Unaligned checkpoints will support rescaling. 
But it will be limited to non-iterative jobs.
This PR adds:
 - validation: non-iterative or non-UC
 - a flag to skip validation
 - passing a flag to subtasks (whether the job is iterative)

## Brief change log

 - Disallow Unaligned Checkpoints for iterative jobs by default
 - Add an option to force Unaligned Checkpoints
 - Add StreamConfig.isGraphContainingLoops

## Verifying this change

This change is a trivial rework without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java [hunk 1]
```diff
@@ -159,4 +159,12 @@
 					"If during checkpointing, checkpoint start delay exceeds this timeout, alignment " +
 					"will timeout and checkpoint barrier will start working as unaligned checkpoint.")
 				.build());
+
+	public static final ConfigOption<Boolean> FORCE_UNALIGNED =
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:163]: You probably need to generate docs and add them to this commit.
` mvn install -Dgenerate-config-docs -pl flink-docs`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 2]
```diff
@@ -524,6 +524,28 @@ public boolean isForceCheckpointing() {
...
+	 * Returns whether Unaligned Checkpoints are force-enabled.
+	 *
+	 * @deprecated Forcing Unaligned Checkpoints will be removed in future version.
+	 */
+	@Deprecated
+	@PublicEvolving
+	public boolean isUnalignedCheckpointsEnabled() {
+		return checkpointCfg.isUnalignedCheckpointsEnabled();
+	}
+
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:537]: The first getter looks odd. At least the description and tags are wrong.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 3]
```diff
@@ -524,6 +524,28 @@ public boolean isForceCheckpointing() {
...
+	/**
+	 * Returns whether Unaligned Checkpoints are force-enabled.
+	 *
+	 * @deprecated Forcing checkpoints will be removed in future version.
+	 */
+	@Deprecated
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:543]: I'd not deprecate it at this point. I see it decoupled from the deprecation of forcing checkpoints.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java [hunk 4]
```diff
@@ -663,6 +664,14 @@ public boolean shouldSortInputs() {
 		return config.get(SORTED_INPUTS);
 	}
 
+	public void setGraphContainingLoops(boolean graphContainingLoops) {
+		config.setBoolean(GRAPH_CONTAINING_LOOPS, graphContainingLoops);
+	}
+
+	public boolean isGraphContainingLoops() {
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:671]: Just to clarify: this is intended for failing late after the user has forced UC and then rescales, right?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 5]
```diff
@@ -216,6 +216,9 @@ private void preValidate() {
 						+ "State checkpoints happen normally, but records in-transit during the snapshot will be lost upon failure. "
 						+ "\nThe user can force enable state checkpoints with the reduced guarantees by calling: env.enableCheckpointing(interval,true)");
 			}
+			if (streamGraph.isIterative() && !checkpointConfig.isForceUnalignedCheckpoints()) {
+				throw new UnsupportedOperationException("Unaligned Checkpoints are currently not supported for iterative jobs");
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:220]: I'd add more details similar to forcing checkpoints

> "Unaligned Checkpoints are currently not supported for iterative jobs, as Flink cannot rescale from these checkpoints. "
						+ "\nYou can force enable unaligned checkpoints if you don't intend to rescale: env.setForceUnalignedCheckpoints(true)"

**Review by AHeise — COMMENTED** [review]
Thank you for your contribution. It looks good % some minor issues.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java (hunk 4 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:671]: No, this is needed to disable restoring channel state for iterative jobs (so they run normally without restoring or failing).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:163]: Done.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 5 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: Done.

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala [hunk 6]
```diff
@@ -881,6 +881,16 @@ class StreamExecutionEnvironment(javaEnv: JavaEnv) {
...
+
+  /**
+   * Returns whether Unaligned Checkpoints are enabled.
+   */
+  def isUnalignedCheckpointsEnabled: Boolean = javaEnv.isUnalignedCheckpointsEnabled
+
+  /**
+   * Returns whether Unaligned Checkpoints are force-enabled.
+   */
+  def isForceUnalignedCheckpoints: Boolean = javaEnv.isForceUnalignedCheckpoints
```
Comment (AHeise) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala:893]: just noticed that you probably also need to adjust `flink-python/pyflink/datastream/checkpoint_config.py`

Comment (rkhachatryan) [reply flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala:893]: Thanks, updated `checkpoint_config.py` too.

**Comment by AHeise** [comment]
There seems to be some consistency check in Python (good job whoever did that)

```
    @classmethod
    def check_methods(cls):
        java_primary_methods = {'getClass', 'notifyAll', 'equals', 'hashCode', 'toString',
                                'notify', 'wait'}
        java_methods = PythonAPICompletenessTestCase.get_java_class_methods(cls.java_class())
        python_methods = cls.get_python_class_methods(cls.python_class())
        missing_methods = java_methods - python_methods - cls.excluded_methods() \
            - java_primary_methods
        if len(missing_methods) > 0:
>           raise Exception('Methods: %s in Java class %s have not been added in Python class %s.'
                            % (missing_methods, cls.java_class(), cls.python_class()))
E           Exception: Methods: {'isForceUnalignedCheckpoints', 'isUnalignedCheckpointsEnabled'} in Java class org.apache.flink.streaming.api.environment.StreamExecutionEnvironment have not been added in Python class <class 'pyflink.datastream.stream_execution_environment.StreamExecutionEnvironment'>.
```

**Comment by rkhachatryan** [comment]
Both CI failures are unrelated:
- TemporalJoinITCase.testEventTimeTemporalJoinChangelogUsingBeforeTime (FLINK-19914)
- SQLClientHBaseITCase (likelly FLINK-19863)

**Pull request merged**